biz.aQute.bndlib/src/aQute/bnd/build/ProjectTracker.java

```diff
@@ -1,95 +1,117 @@
 package aQute.bnd.build;
 
-import java.io.File;
-import java.util.ArrayList;
+import static java.nio.file.StandardWatchEventKinds.ENTRY_CREATE;
+import static java.nio.file.StandardWatchEventKinds.ENTRY_DELETE;
+
+import java.io.IOException;
+import java.nio.file.ClosedWatchServiceException;
+import java.nio.file.DirectoryStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.WatchKey;
+import java.nio.file.WatchService;
 import java.util.HashMap;
 import java.util.HashSet;
-import java.util.List;
 import java.util.Map;
 import java.util.Optional;
 import java.util.Set;
 
+import aQute.lib.exceptions.Exceptions;
 import aQute.lib.io.IO;
 
 /**
  * This class is responsible for maintaining the project list. Since this list
  * can change asynchronously we guard access from here.
  */
-class ProjectTracker {
-	final Map<String, Project>	models		= new HashMap<>();
-	final List<Project>			buildOrder	= new ArrayList<>();
+class ProjectTracker implements AutoCloseable {
+	final Map<String, Project>	models	= new HashMap<>();
 	final Workspace				workspace;
+	final Path					base;
+	final WatchService			watchService;
+	final WatchKey				watchKey;
 
-	boolean						inited		= false;
+	boolean						changed	= true;
 
-	ProjectTracker(Workspace workspace) {
+	ProjectTracker(Workspace workspace) throws IOException {
 		this.workspace = workspace;
+		base = workspace.getBase()
+			.toPath();
+		watchService = base.getFileSystem()
+			.newWatchService();
+		watchKey = base.register(watchService, ENTRY_CREATE, ENTRY_DELETE);
+	}
+
+	@Override
+	public synchronized void close() {
+		IO.close(watchService);
+		models.values()
+			.forEach(IO::close);
 	}
 
 	/*
 	 * Next time the list must be refreshed
 	 */
 	synchronized void refresh() {
-		inited = false;
+		changed = true;
 	}
 
 	/*
 	 * Answer a snapshot of the current list of projects
 	 */
 	synchronized Set<Project> getAllProjects() {
-		init();
+		update();
 		return new HashSet<>(models.values());
 	}
 
 	/*
 	 * Answer the project with the given name
 	 */
 	synchronized Optional<Project> getProject(String name) {
-		init();
+		update();
 		return Optional.ofNullable(models.get(name));
 	}
 
 	/*
 	 * Sync the directories with the current lists of projects
 	 */
-	private void init() {
-		if (inited)
-			return;
+	private void update() {
+		try {
```

Comment:
I am not happy with this change. 
a) This code is unnecessary for batch  builds so it should be implemented in the driver, the API of Workspace has a refreshProjects() method for this. 
b) I have terrible experiences with WatchService in MacOS and Windows 7. Might be solved now, but still
c) It creates an indeterminism which makes it impossible to rely on since the even is delivered async (and on Mac could take seconds)
d) If we want to support watching directory we should centralize it and not use the raw API in this class. I.e. Workspace could have a watch( File, Callback ) method to subscribe to changes on the file system that can then share the same service.